File: LeetCode/four_sum.py
def four_sum(nums: List[int], target: int) -> List[List[int]]:
    nums.sort()
    result = []

    for i in range(len(nums) - 3):
        # Skip duplicates for the first number
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        for j in range(i + 1, len(nums) - 2):
            # Skip duplicates for the second number
            if j > i + 1 and nums[j] == nums[j - 1]:
                continue

            left, right = j + 1, len(nums) - 1

            while left < right:
                total = nums[i] + nums[j] + nums[left] + nums[right]

                if total == target:  
                    result.append([nums[i], nums[j], nums[left], nums[right]])

                    # Skip duplicates for the third and fourth numbers
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1

                    left += 1
                    right -= 1

                elif total < target:
                    left += 1
                else:
                    right -= 1

    return result

File: LeetCode/solve_in_rust/src/four_sum.rs
impl Solution {
    pub fn four_sum(nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut nums = nums;
        nums.sort();
        let mut res: Vec<Vec<i32>> = Vec::new();
        if nums.len() < 4 {
            return res;
        }

        for i in 0..nums.len() - 3 {
            if i > 0 && nums[i] == nums[i - 1] {
                continue;
            }
            for j in i + 1..nums.len() - 2 {
                if j > i + 1 && j < nums.len() - 2 && nums[j] == nums[j - 1] {
                    continue;
                }
                let mut left = j + 1;
                let mut right = nums.len() - 1;

                while left < right {
                    let sum = nums[i] + nums[j] + nums[right] + nums[left];
                    if sum > i32::MAX {
                        continue;
                    }
                    if sum == target {
                        let temp = vec![nums[i], nums[j], nums[left], nums[right]];
                        res.push(temp);

                        right -= 1;
                        left += 1;
                        while left < right && nums[left] == nums[left - 1] {
                            left += 1;
                        }

                        while right > left && nums[right] == nums[right + 1] {
                            right -= 1;
                        }
                    } else if sum > target {
                        right -= 1;
                    } else if sum < target {
                        left += 1;
                    }
                }
            }
        }
        return res;
    }
}
